geomstats/geometry/siegel.py

```diff
@@ -0,0 +1,614 @@
+    https://epubs.siam.org/doi/pdf/10.1137/15M102112X
+"""
+
+import geomstats.backend as gs
+from geomstats.geometry.base import ComplexOpenSet
+from geomstats.geometry.complex_matrices import ComplexMatrices
+from geomstats.geometry.complex_riemannian_metric import ComplexRiemannianMetric
+from geomstats.geometry.hermitian_matrices import HermitianMatrices
+
+
+def create_identity_mat(shape, dtype=gs.get_default_cdtype()):
+    """Stack identity matrices.
+
+    Parameters
+    ----------
+    shape : tuple
+        Desired identity matrix shape of form [..., n, n].
+    dtype : dtype
+        Desired dtype.
+        Optional, default: gs.get_default_cdtype().
+
+    Returns
+    -------
+    identity : array-like, shape=[..., n, n]
+        Stacked identity matrices.
+    """
+    ndim = len(shape)
+    if ndim == 2:
+        identity = gs.eye(shape[-1], dtype=dtype)
+    elif ndim == 3:
+        identity = gs.stack(
+            [gs.eye(shape[-1], dtype=dtype) for _ in range(shape[0])], axis=0
+        )
+    return identity
+
+
+class Siegel(ComplexOpenSet):
+    """Class for the Siegel space.
+
+    Parameters
+    ----------
+    n : int
+        Integer representing the shape of the matrices: n x n.
+    symmetric : bool
+        If symmetric is True, add a symmetry condition
+        on the matrices to belong to the Siegel space.
+        Optional, default: False.
+    scale : float
+        Scale of the complex Poincare metric.
+        Optional, default: 1.
+    """
+
+    def __init__(self, n, symmetric=False, scale=1.0, **kwargs):
+        kwargs.setdefault("metric", SiegelMetric(n))
+        super().__init__(
+            dim=n**2, embedding_space=ComplexMatrices(m=n, n=n), **kwargs
+        )
+        self.n = n
+        self.symmetric = symmetric
+        self.scale = scale
+
+    def belongs(self, point, atol=gs.atol):
+        """Check if a matrix belongs to the Siegel space.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., n, n]
+            Point to be checked.
+        atol : float
+            Tolerance.
+            Optional, default: backend atol.
+
+        Returns
+        -------
+        belongs : array-like, shape=[...,]
+            Boolean denoting if mat belongs to the Siegel space.
+        """
+        point_shape = gs.shape(point)
+
+        ndim = len(point_shape)
+        if ndim == 3:
+            n_samples = point_shape[0]
+
+        point_transconj = ComplexMatrices.transconjugate(point)
+        aux = gs.einsum("...ij,...jk->...ik", point, point_transconj)
+
+        if ndim == 2:
+            eigenvalues = gs.linalg.eigvalsh(aux)
+            belongs = gs.all(eigenvalues <= 1 - atol)
+        elif ndim == 3:
+            belongs = []
+            for i_sample in range(n_samples):
+                eigenvalues = gs.linalg.eigvalsh(aux[i_sample, ...])
+                belongs.append(gs.all(eigenvalues <= 1 - atol))
+            belongs = gs.array(belongs)
```

Comment:
no need for loop here: `eigvalsh` is vectorized.

Reply:
Done.